### PR TITLE
Add aws_default_network_acl, aws_network_acl and aws_network_acl_rule

### PR DESCRIPTION
**Pull request by eliecharra**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | yes
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | #580 
| ❓ Documentation  | https://github.com/cloudskiff/driftctl-docs/pull/121

### CODE REVIEW

File: pkg/driftctl.go [hunk 1]
```diff
@@ -111,6 +112,8 @@ func (d DriftCTL) Run() (*analyser.Analysis, error) {
 	if !d.opts.StrictMode {
 		middleware = append(middleware,
 			middlewares.NewAwsDefaults(),
+			middlewares.NewAwsDefaultNetworkACL(),
+			middlewares.NewAwsDefaultNetworkACLRule(),
```
Comment (wbeuil) [inline pkg/driftctl.go:116]: Don't understand why you would want these two middlewares to be in the strict-mode.

If you want to manage and see diffs on your default network acl you'd need to enable strict-mode and deep-mode.

Why not run both like the others (e.g. default-vpc, default-rout-table ...) ?

File: pkg/middlewares/aws_default_network_acl_rule.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	}
+
+	if cidr := res.Attrs.GetString("cidr_block"); cidr != nil && *cidr != "0.0.0.0/0" {
+		return false
+	}
+
+	if proto := res.Attrs.GetString("protocol"); proto != nil && *proto != "-1" {
+		return false
+	}
+
```
Comment (wbeuil) [inline pkg/middlewares/aws_default_network_acl_rule.go:79]: It misses a case with ipv6_cidr_block. Test with this tf file:

```
resource "aws_vpc" "foo" {
    cidr_block = "10.1.0.0/16"
    assign_generated_ipv6_cidr_block = true
}

resource "aws_network_acl" "bar" {
  vpc_id = aws_vpc.foo.id
}

resource "aws_network_acl_rule" "bar" {
  network_acl_id = aws_network_acl.bar.id
  rule_number    = 200
  egress         = false
  protocol       = "tcp"
  rule_action    = "allow"
  # cidr_block     = aws_vpc.foo.cidr_block
  ipv6_cidr_block = aws_vpc.foo.ipv6_cidr_block
  from_port      = 22
  to_port        = 22
}
```

**Review by wbeuil — CHANGES_REQUESTED** [review]
Remind me of security group rules but with more edge cases ^^

File: pkg/remote/aws/repository/ec2_repository.go [hunk 3]
```diff
@@ -328,3 +335,30 @@ func (r *ec2Repository) ListAllSecurityGroups() ([]*ec2.SecurityGroup, []*ec2.Se
 	r.cache.Put("ec2ListAllDefaultSecurityGroups", defaultSecurityGroups)
 	return securityGroups, defaultSecurityGroups, nil
 }
+
+func (r *ec2Repository) ListAllNetworkACLs() ([]*ec2.NetworkAcl, error) {
+
+	cacheKey := "ec2ListAllNetworkACLs"
+	r.locks[cacheKey].Lock()
+	defer r.locks[cacheKey].Unlock()
```
[the remark this answers is not in the document]
Comment (sundowndev) [reply pkg/remote/aws/repository/ec2_repository.go:343]: not sure if it's a good idea to implement an independent solution for #963 in this PR, I think we need to agree on a common solution for all repositories and providers. WDYT?

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/remote/aws/repository/ec2_repository.go:343]: > From this lock, I understand that the lock inside the r.cache.Get is not enough if the method ListAllNetworkACLs gets called twice in our ParallelRunner. Is that the case?

No, since if we got a cache miss we only want to release the lock after we read value from AWS. If we do not do that you can have two parallel cache miss, we have opened an issue for that : #963 
 
> not sure if it's a good idea to implement an independent solution for #963 in this PR, I think we need to agree on a common solution for all repositories and providers. WDYT?

The idea was not to implement a solution for #963 here but only a single required lock for this repository. Do you think that we should ship this PR without any lock knowing that it will be broken ?

File: pkg/remote/aws/ec2_network_acl_rule_enumerator.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+			results = append(
+				results,
+				e.factory.CreateAbstractResource(
+					string(e.SupportedType()),
+					"",
+					map[string]interface{}{
+						"egress":         *entry.Egress,
+						"network_acl_id": *res.NetworkAclId,
+						"rule_action":    *entry.RuleAction,          // Used in default middleware
+						"cidr_block":     *entry.CidrBlock,           // Used in default middleware
```
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/remote/aws/ec2_network_acl_rule_enumerator.go:45]: I should double check the ipv6 case and add a test case for it, I'm not sure that we'll retrieve a nil value or an empty string here. Anyway this should be investigated and covered by tests nice catch thanks 👍🏻

File: pkg/remote/aws/ec2_network_acl_enumerator.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+		return nil, remoteerror.NewResourceListingError(err, string(e.SupportedType()))
+	}
+
+	results := make([]*resource.Resource, 0, len(resources))
+
+	for _, res := range resources {
+		// Do not handle default network acl since it is a dedicated resource
+		if *res.IsDefault {
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/remote/aws/ec2_network_acl_enumerator.go:38]: Yeah I think it was a mistake to make `ListAllVPCs` aware of the default concept, this is not the role of a repository. IMHO  BTW the way we use two cache key in this method is weird. There is only one thing that we want to cache : aws sdk call result, it does not make a lot of sense to cache these two results. I mean the cost of iterating the slice and filter default vpc is insignificant compared to a network call ... I also like the idea to have a single enumerator per resource type, the way we did VPC was an anti-pattern for me, maybe we do it like this before we introduced the repository pattern and it was to avoid doing the API call twice ?

File: pkg/remote/aws/repository/ec2_repository.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sundowndev) [reply pkg/remote/aws/repository/ec2_repository.go:343]: > Do you think that we should ship this PR without any lock knowing that it will be broken ?

We already have tons of resources that misses a locking system anyway, may be #963 should be considered as a release blocker then?

File: pkg/driftctl.go (hunk 1 above)
Comment (eliecharra) [reply pkg/driftctl.go:116]: Yeah you are right I messed up with strict mode scope in my mind thanks 👍🏻

File: pkg/remote/aws/ec2_network_acl_enumerator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (wbeuil) [reply pkg/remote/aws/ec2_network_acl_enumerator.go:38]: Yep we did create the repository pattern after. Ok I'm good with your way of only having one enumerator for 2 "identical" resources (e.g. default and non-default)

File: pkg/remote/aws/repository/ec2_repository.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (wbeuil) [reply pkg/remote/aws/repository/ec2_repository.go:343]: I'm not against fixing this in newer resources but let's be clear right now before merging new ones.

File: pkg/middlewares/aws_default_network_acl_rule.go (hunk 2 above)
Comment (eliecharra) [reply pkg/middlewares/aws_default_network_acl_rule.go:79]: I don't think this is relevant here, actually there is no default IPv6 rules

File: pkg/remote/aws/repository/ec2_repository.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sundowndev) [reply pkg/remote/aws/repository/ec2_repository.go:343]: > There is a difference between refactoring legacy code to fix technical debt and shipping new code with known technical debt

I agree with that, I just think it's better to solve the problem once for all. May be we could ensure the method locks as expected in tests for new resources, so we avoid any regression when we will work on #963 ?

File: pkg/middlewares/aws_default_network_acl_rule.go (hunk 2 above)
Comment (wbeuil) [reply pkg/middlewares/aws_default_network_acl_rule.go:79]: Are you talking about the default one on each region or the default one created once you create a new VPC ?

File: pkg/remote/aws/repository/ec2_repository.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (moadibfr) [reply pkg/remote/aws/repository/ec2_repository.go:343]: I wonder if we could have a solution where the cache contains the lock and we call it the same way from all the repository.
something like 
```go
val, lock := r.cache.Read(cacheKey);
if  val != nil {
   return val.([]*ec2.NetworkAcl), nil
}
defer lock..Unlock()
```
And cache would lock if there is a miss and pass the responsability for the unlock ?

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/remote/aws/repository/ec2_repository.go:343]: We can debate about implementation details later. The main question for this PR is : should we handle lock in future repositories or should we wait for #963 to fix them all in a second time ?

[the remark this answers is not in the document]
Comment (moadibfr) [reply pkg/remote/aws/repository/ec2_repository.go:343]: If we want a common solution for all I think we should not fix it in new resources I think.
But if we want each repo to handle it's own lock (which makes sense) we could fix in new resources and take care of the old ones later

[the remark this answers is not in the document]
Comment (eliecharra) [reply pkg/remote/aws/repository/ec2_repository.go]: I removed locks for now, this PR is big enough, we'll see how to fix this in #963. Maybe we should prioritize it.

**Review by wbeuil — APPROVED** [review]

File: pkg/middlewares/aws_default_network_acl_rule.go (hunk 2 above)
Comment (eliecharra) [reply pkg/middlewares/aws_default_network_acl_rule.go:91]: I added ipv6 handling here you are right
